src/engraving/dom/arpeggio.cpp

```diff
@@ -52,13 +54,130 @@ Arpeggio::Arpeggio(Chord* parent)
     m_userLen2 = 0.0;
     m_playArpeggio = true;
     m_stretch = 1.0;
+
+    parent->setSpanArpeggio(this);
+}
+
+Arpeggio::~Arpeggio()
+{
+    // Remove reference to this arpeggio in any chords it may have spanned
+    removeChords(track(), track() + m_span);
 }
 
 const TranslatableString& Arpeggio::arpeggioTypeName() const
 {
     return TConv::userName(m_arpeggioType);
 }
 
+void Arpeggio::findChords()
+{
+    Chord* _chord = chord();
+    track_idx_t strack = track();
+    track_idx_t etrack = track() + (m_span - 1);
+
+    for (track_idx_t track = strack; track <= etrack; track++) {
+        EngravingItem* e = _chord->segment()->element(track);
+        if (e && e->isChord()) {
+            toChord(e)->setSpanArpeggio(this);
+        }
+    }
+}
+
+void Arpeggio::removeChords(track_idx_t strack, track_idx_t etrack)
+{
+    Chord* _chord = chord();
+    for (track_idx_t track = strack; track <= etrack; track++) {
+        EngravingItem* e = _chord->segment()->element(track);
+        if (e && e->isChord() && toChord(e)->spanArpeggio() == this) {
+            toChord(e)->setSpanArpeggio(nullptr);
+        }
+    }
+}
+
+void Arpeggio::rebaseStartAnchor(int direction)
+{
+    if (direction == 0) {
+        return;
+    }
+    if (direction == 1) {
+        // Move arpeggio to chord above
+        Staff* s = staff();
+        Part* part = s->part();
+        staff_idx_t topStaff = *part->staveIdxList().begin();
+        track_idx_t topTrack = topStaff * VOICES;
+        if (track() > topTrack) {
+            // Loop through voices til we find a chord
+            for (track_idx_t curTrack = track() - 1; curTrack >= topTrack; curTrack--) {
+                EngravingItem* e = chord()->segment()->element(curTrack);
+                if (e && e->isChord()) {
+                    int newSpan = m_span + track() - curTrack;
+                    if (newSpan != 0) {
+                        undoChangeProperty(Pid::ARPEGGIO_SPAN, newSpan);
+                        score()->undo(new ChangeParent(this, e, e->staffIdx()));
+                        break;
+                    }
+                }
+            }
+        }
+    } else if (direction == -1) {
+        // Move arpeggio to chord below
+        for (track_idx_t curTrack = track() + 1; curTrack <= track() + m_span - 1; curTrack++) {
+            EngravingItem* e = chord()->segment()->element(curTrack);
+            if (e && e->isChord()) {
+                int newSpan = m_span + track() - curTrack;
+                if (newSpan != 0) {
+                    chord()->setSpanArpeggio(nullptr);
```

Comment:
Is it okay that this is done in a non-undoable way?

Reply:
I doubt it!  It also definitely wants to be propagated to parts